src/protocols/bdx/BdxTransferSession.h

```diff
@@ -347,11 +342,51 @@ class DLL_EXPORT TransferSession
      */
     CHIP_ERROR VerifyProposedMode(const BitFlags<TransferControlFlags> & proposed);
 
+    void SendStatusReportWithError(StatusReportData statusReportData);
     void PrepareStatusReport(StatusCode code);
     bool IsTransferLengthDefinite() const;
 
+    template <typename MessageType>
+    void PrepareOutgoingMessageEvent(MessageType messageType);
+
+    /**
+     * @brief
+     *   Register a callback with the transfer session. The callback will be called to notify the caller about any messages received
+     *   or messages that need to be send or errors that occurred internally.
+     *
+     * @param callback    OutputEventCallback that is to be registered with the transfer session and gets called when Output events
+     *                    are generated.
+     * @param context     Reference to the caller of this method
+     */
+    void RegisterOutputEventCallback(OutputEventCallback callback, void * context);
+
+    /**
+     * @brief
+     *   Unregister the OutputEventCallback previously registered with the transfer session.
+     *
+     */
+    void UnregisterOutputEventCallback();
+
+    /**
+     * @brief
+     *   Sends the generated output event from the transfer session - BDX messages, status reports, etc to
+     *   the consumers of the messages who register the OutputEventCallback with it.
+     */
+    void SendOutputEvent(chip::bdx::TransferSession::OutputEvent & outputEvent);
+
+    /**
+     * The context that is passed in along with the callback. When the output event is generated, the context is passed when the
+     * OutputEventCallback is called.
```

Comment:
```suggestion
     * The context that is passed to the registered OutputEventCalllback.
```
(or OutputEventHandler, whatever the naming).